Add CreateTxError and use as error type for TxBuilder::finish()

notmandatory opened the pull request:
### Description

To remove some places where there were `.expect("TODO")` I added a new `CreateTxError` type which is returned from `TxBuilder::finish()`. I also updated related tests and doc tests.

Fixes https://github.com/bitcoindevkit/bdk/issues/996#issuecomment-1621036206

Also added fallible `Wallet::try_get_address()` and `try_get_internal_address()`  to return `Result` with a possible `D:WriteError` when a PersistBackend is used. This should fix bitcoindevkit/bdk#996.

I removed catch-all bdk::Error and replaced usages with new types and updated related functions, fixes bitcoindevkit/bdk#994.

### Notes to the reviewers

~~I didn't add all possible bdk::Error types that `Wallet::create_tx()` and `TxBuilder::finish()` functions might throw. It's probably not too much more work but will take a bit more research so I want to make sure this is the right general approach first.~~

I added `anyhow` to the dev-dependencies so I could remove some `.expect()` lines from the docs tests and make the examples closer to what an end user should do.  I also used the `anyhow!()` macro to replace a few places that were using the `bdk::Error::Generic` in example code.

I also moved the module level error.rs file to wallet/error.rs so no one would be tempted to make any new catch all errors and to make it clear that all the errors in it are wallet module related. 

### Changelog notice

Changed

- Updated bdk module to use new context specific error types
  - wallet: MiniscriptPsbtError, CreateTxError, BuildFeeBumpError error enums
  - coin_selection: module Error enum
- Renamed fallible Wallet address functions to try_get_address() and try_get_internal_address()

Removed

- Removed catch-all top level bdk::Error enum
- Removed impl_error macro

Added

-  Added infallible Wallet get_address(), get_internal_address functions

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### Bugfixes:

* [x] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [ ] I'm linking the issue being fixed by this PR

evanlinjin commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -200,6 +201,61 @@ pub enum InsertTxError {
+    Persist(P),
+    /// There was a problem while extracting and manipulating policies
+    Policy(PolicyError),
+    /// TODO: replace this with specific error types
```
I think it will be a good idea to remove the `CreateTxError::Bdk` branch in this PR as well. I think removing this will improve the API significantly (since the caller won't have to handle a large list of redundant error variants).

This can be our opportunity to avoid using `Error::Generic` and have "concrete" variants for timelock/tx-version errors.

If needed, we can have `CreateTxError` be a variant of `bdk::Error`, although I would avoid this if possible.

notmandatory replied:
If you like the general approach of adding a new `CreateTxError` enum but would rather that I replace the `Bbk` variant with the specific errors for this PR then I'm happy to do that. I just wanted to make sure I was on the right track before chasing down all the required variants.

notmandatory replied:
I should be able to get this knocked out today or tomorrow so we don't delay the alpha.1 too much more.

notmandatory replied:
So this took a bit longer than a day or two to finally wrap up, but should be all done now. 😅

evanlinjin reviewed (changes requested):
Thank you for taking this on!

I appreciate your attempt to be conservative in the amount of changes made. However, the current state of this PR does not result in a better API since the caller still has to handle a large list of redundant error variants.

If it was a PR to purely remove the `.expect("TODO")`s, I would just return a `bdk::Error::Generic` or introduce a new variant in `bdk::Error`.

However, I would prefer this PR to fix the error-handling API for tx creation completely (as it seems you already have done the majority of the work).

evanlinjin commented:
But we probably just need a quick fix for the next release.

notmandatory commented:
I added fallible `Wallet::try_get_address()` and `try_get_internal_address()`  to return `Result` with a possible `D:WriteError` when a PersistBackend is used. This should fix bitcoindevkit/bdk#996.

notmandatory commented:
I removed the `CreateTxError::Bdk` variant and moved all the needed variants from the `bdk::Error` to the `CreateTxError` enum, plus created concrete variants for any errors that were using `Generic`. A few other things I'd like to do, but don't need to be in this PR:

* add `BuildFeeBumpError` enum for `Wallet::build_fee_bump` 
* add `SignError` enum for `Wallet::sign`
* add `FinalizeError` enum for `Wallet::finalize` 
* eliminate (or greatly diminish) `bdk::Error`
* move the wallet error enums and related code to a wallet::error file
* add fallible and infallible  versions of any others Wallet functions it makes sense for

notmandatory commented:
Per comment: https://github.com/bitcoindevkit/bdk/issues/1025#issuecomment-1635184287

I'm moving this to future alpha release.

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -159,14 +160,32 @@ impl Wallet {
+    /// (i.e. does not end with /*) then the same address will always be returned for any [`AddressIndex`].
+    pub fn get_address(&mut self, address_index: AddressIndex) -> AddressInfo {
+        self.try_get_address(address_index).unwrap()
+    }
```
This is implemented on `Wallet<()>` but it should be implemented on `Wallet<D> where D: PersistBackend<WriteError=core::convert::Infallible>`

notmandatory replied:
If I do this it looks like I also have to add a `db: D` parameter to `new_no_persist()`, and then all callers (tests for now) need to specify a `db: ()`. Is there scenario you're thinking of where D is some other `Infallible` database?

notmandatory replied:
@LLFourn I'd like to wrap up this PR, but want to make sure you're OK leaving this as it is. I don't see any reason to pass in a `db:D` parameter for `new_no_persist()` when it'd always be `()`.

danielabrozzoni replied:
I think Lloyd meant that you should keep `new_no_persist` in `impl Wallet`, but move the other two methods in `impl Wallet<D: ...>`:
```diff
diff --git a/crates/bdk/src/wallet/mod.rs b/crates/bdk/src/wallet/mod.rs
index 051f9730..061a3db5 100644
--- a/crates/bdk/src/wallet/mod.rs
+++ b/crates/bdk/src/wallet/mod.rs
@@ -226,7 +226,12 @@ impl Wallet {
             NewError::Persist(_) => unreachable!("no persistence so it can't fail"),
         })
     }
+}
 
+impl<D> Wallet<D>
+where
+    D: PersistBackend<ChangeSet, WriteError = core::convert::Infallible>,
+{
     /// Infallibly return a derived address using the external descriptor, see [`AddressIndex`] for
     /// available address index selection strategies. If none of the keys in the descriptor are derivable
     /// (i.e. does not end with /*) then the same address will always be returned for any [`AddressIndex`].

```

To me, it makes sense to have `get_address` and `get_internal_address` for all the wallets that have infallible backends (which could be either `()` or some other database)

notmandatory replied:
Fixed!

LLFourn commented on example-crates/wallet_electrum/src/main.rs:
```diff
@@ -26,7 +26,9 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
-    let address = wallet.get_address(bdk::wallet::AddressIndex::New);
+    let address = wallet
+        .try_get_address(bdk::wallet::AddressIndex::New)
+        .expect("new_address");
```
You should not panic when this fails. It indicates that persistence failed. A good way to do this might be to use `anyhow::Context::context` and then `?` the error. It doesn't look like the examples depend on `anyhow` but they should and the main function should return `anyohw::Result`.

notmandatory replied:
I added `anyhow` and removed the `.expect()` but didn't add the `.context()` because I don't think it's needed for this simple example.

LLFourn reviewed (dismissed):
+1 on adding `CreateTxError`. left some comments.

nondiremanuel commented:
@notmandatory we didn't talk about this PR in the last calls. Is it correct to keep it into alpha.3 milestone? Shall we discuss it in the next call?

notmandatory commented:
> @notmandatory we didn't talk about this PR in the last calls. Is it correct to keep it into alpha.3 milestone? Shall we discuss it in the next call?

I don't think we need to discuss it until after alpha.2 is released. Then I'll have to rebase it and fix a few small things and will bug folks for a re-review.

notmandatory commented:
I've rebased and now need to do fixes suggested in comments.

notmandatory commented:
This should also fix bitcoindevkit/bdk#994.  I'm in the process now of removing wallet::Error enum and replacing with specific errors. 

~~I'm also adding `thiserror` crate for less boilerplate error code.~~ We can't use `thiserror` because it isn't supported with no_std. :-(

notmandatory commented:
Moved this back to draft since still have more work to do before ready for a re-review.

notmandatory commented:
This is ready for a re-review with the caveat that after CI fixes are made on `master` branch I'll need to rebase and remove the CI related commits I added here.

realeinherjar commented on .github/workflows/cont_integration.yml:
```diff
@@ -12,7 +12,7 @@ jobs:
           - version: stable
             clippy: true
-          - version: 1.57.0 # MSRV
+          - version: 1.63.0 # MSRV
```
Just a heads-up and to link the PRs.
This depends on #1183

notmandatory replied:
Yup I'll remove these commits and rebase on #1183 after it's merged.

realeinherjar commented on crates/bdk/Cargo.toml:
```diff
@@ -45,8 +45,9 @@ dev-getrandom-wasm = ["getrandom/js"]
 [dev-dependencies]
 lazy_static = "1.4"
-env_logger = "0.7"
+env_logger = { version = "0.7", default-features = false }
```
I am a little bit picky with changes to `Cargo.toml`.
I understand the `anyhow` addition, but why changing `env_logger` to `default-features = false`?
And I could not find this documented in the PR.

notmandatory replied:
I originally changed this to remove one additional MSRV pinning, but was probably being a bit to aggressive so I put it back as it was. I also put it back in #1183.

realeinherjar commented on crates/bdk/src/descriptor/error.rs:
```diff
@@ -10,7 +10,7 @@
 
 //! Descriptor errors
-
+#[cfg(feature = "std")]
```
What is the impact for `no-std`?
What would be the returned errors in `no-std`?

notmandatory replied:
Good question, I config'd out `fmt::Display` for errors here for consistence since it was done in other parts of the bdk package. Technically I think we could leave it and it wouldn't break `no_std` code but I believe it would make the final binary larger which could be a problem for some `no_std` embedded devices.  

@afilini any thoughts on if we should or should not include `fmt::Display` impls for errors with `no_std`?

evanlinjin replied:
I don't think it will hurt to impl Display for `no_std`. Also curious on the reasoning.

notmandatory replied:
To answer @realeinherjar there would still be returned error enums but just not formatted .to_string() or in a display string. 

I also found this list of suggestions for making small `no_std` binaries, including the recommendation to not use `core::fmt`. Looks like it only applies if someone is using `bdk` and needs an executable smaller than 20 kilobytes.

[Remove core::fmt with #![no_main] and Careful Usage of libstd](https://github.com/johnthagen/min-sized-rust#remove-corefmt-with-no_main-and-careful-usage-of-libstd)

notmandatory replied:
I removed the `no_std` cfgs I added on `fmt:Display`. This change would have to be done for all bdk crates for it to have any effect on `no_std` binary size and I don't think are blocked by this. I created a new issue #1198 targeted for the 2.0 milestone.

notmandatory commented:
I've removed the cfg for fmt::Display and created a new issue for that change (if we want to do it). Tests passing and ready for re-review.

ValuedMammal commented on crates/bdk/examples/mnemonic_to_descriptors.rs:
unrelated to this PR, but the derivation path should correspond to the network, e.g. `m/86h/1h/0h/0`

notmandatory replied:
I created #1209 to check this in all our docs and examples.

ValuedMammal commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -1396,15 +1442,16 @@ impl<D> Wallet<D> {
     /// };
     /// let  finalized = wallet.sign(&mut psbt, SignOptions::default())?;
```
nit: whitespace
```suggestion
    /// let finalized = wallet.sign(&mut psbt, SignOptions::default())?;
```

ValuedMammal commented on crates/bdk/src/wallet/tx_builder.rs:
```diff
@@ -540,7 +554,7 @@ impl<'a, D, Cs: CoinSelectionAlgorithm, Ctx: TxBuilderContext> TxBuilder<'a, D,
     /// Returns the [`BIP174`] "PSBT" and summary details about the transaction.
```
```suggestion
    /// Returns a new [`Psbt`] per [BIP174].
    ///
    /// [BIP174]: https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki
```

ValuedMammal commented on crates/bdk/src/wallet/error.rs:
```diff
@@ -0,0 +1,407 @@
+impl std::error::Error for AddUtxoError {}
+
+#[derive(Debug)]
+/// Error returned from [`TxBuilder::add_utxo`] and [`TxBuilder::add_utxos`]
```
```suggestion
/// Error returned from [`TxBuilder::add_foreign_utxo`]
```

ValuedMammal commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -1416,7 +1463,7 @@ impl<D> Wallet<D> {
                 .any(|i| i.non_witness_utxo.is_none())
         {
-            return Err(Error::Signer(signer::SignerError::MissingNonWitnessUtxo));
+            return Err(SignError::Signer(SignerError::MissingNonWitnessUtxo));
```
Maybe `Wallet::sign` and `Wallet::finalize_psbt` can return `Result<bool, signer::SignerError>` to avoid having to wrap it in a new type.. but that might require `signer::SignerError` to absorb an extra variant `MiniscriptPsbt`

notmandatory replied:
Good idea! fixed.

ValuedMammal commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -38,6 +38,7 @@ use bitcoin::{
+use descriptor::error::Error as DescriptorError;
 use miniscript::psbt::{PsbtExt, PsbtInputExt, PsbtInputSatisfier};
 
 use bdk_chain::tx_graph::CalculateFeeError;
```
should `CalculateFeeError` also implement std error so we don't have to call `expect` on `Wallet::calculate_fee`?

notmandatory replied:
Makes sense, I added the implementation.

ValuedMammal reviewed (commented):
I made minor comments, but nothing that should block this from moving forward

notmandatory commented:
I pushed a couple more commits to add changes suggested by @ValuedMammal. I'll squash commits to clean up the history prior to merging.

matthiasdebernardini commented:
Hey! This needs fixing too, "not enough" 

https://github.com/bitcoindevkit/bdk/blob/d6a0cf0795d9f8a1550e614e96b339a18fc537dc/crates/bdk/src/descriptor/policy.rs#L524

vladimirfomene commented on crates/bdk/src/wallet/coin_selection.rs:
```diff
@@ -100,14 +100,16 @@
 
 use core::convert::TryInto;
+use core::fmt;
+use core::fmt::Formatter;
```
Nits `use core::fmt::{self, Formatter};`

vladimirfomene commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -67,11 +70,11 @@ use crate::descriptor::{
     calc_checksum, into_wallet_descriptor_checked, DerivedDescriptor, DescriptorMeta,
```
nit: add `self` on the import here to avoid importing on line 61

vladimirfomene commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -224,6 +227,24 @@ impl Wallet {
+        self.try_get_address(address_index).unwrap()
+    }
+
+    /// Infallibly return a derived address using the internal (change) descriptor.
```
I'm curious to know why we needed to create infallible version of the address functions.

notmandatory replied:
I think the only reason for this is so that non-persistent / memory only applications don't need to handle a persistence error which will never happen. This is probably a rare case but I suspect there are some embedded system applications where this would be useful.

evanlinjin replied:
This is an excellent addition, thank you @notmandatory.

vladimirfomene reviewed (approved):
Just nits. LGTM. ACK 35a108b4300473277e6f575a8f781add5d555a48

evanlinjin commented:
I don't think this should be addressed in this PR, but I think we should use idiomatic naming for methods: https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

I.e. `Wallet::get_address` should be `Wallet::address`.

evanlinjin reviewed (approved):
ACK 35a108b4300473277e6f575a8f781add5d555a48

Thank you for this work. Now the errors are more domain-specific and detailed.

danielabrozzoni reviewed (commented):
Fantastic work, than you.

I think Lloyd's comment should be fixed, as it should be fairly trivial. I left a comment on that.

notmandatory commented:
> I think Lloyd's comment should be fixed, as it should be fairly trivial. I left a comment on that.

Ohhhh! 🤦‍♂️ thanks for the help here, now I get it. I'm going to push your fix, rebase, clean up the history a little, and then it should be ready to merge.

notmandatory committed:
refactor(bdk)!: add context specific error types, remove top level error mod

refactor(bdk)!: remove impl_error macro
refactor(wallet)!: add MiniscriptPsbtError, CreateTxError, BuildFeeBumpError error enums
refactor(coin_selection)!: add module Error enum
test(bdk): use anyhow dev-dependency for all tests

notmandatory committed:
feat(wallet): Add infallible Wallet get_address(), get_internal_address functions

refactor!(wallet)!: rename fallible Wallet try_get_address(), try_get_internal_address functions

notmandatory committed:
ci: fix MSRV pinning for rustls 0.21.9
